typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1361 @@
+            relax it to Unique, and we will make it Shared, and make m0
+            Shared for code-motion. However, there is no good way to do that,
+            because the "unique_use" in "maybe_unique" is not complete,
+            because the type-checking and uniqueness analysis is performed on
+            a per-top-level-expr basis.
+
+            Our solution is to record on the m0 that it is constrained by the
+            m1. I.e. if m1 is Unique, then m0 cannot be Shared. After the type
+            checking of the whole file, m1 will correctly tells whether it needs
+            to be Unique, and by extension whether m0 can be Shared. *)
+        let uniq = Maybe_unique.uniqueness l1 in
+        Maybe_shared.set_barrier l0 uniq;
+        m1
+    | Shared _, Borrowed _ -> m0
+    | Maybe_unique l, Borrowed occ ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared _, Maybe_shared _ -> m0
+    | Maybe_unique l0, Maybe_shared l1 ->
+        (* Four cases:
+            Shared;Borrowed = Shared
+            Shared;Shared = Shared
+            Unique;Borrowed = Error
+            Unique;Shared = Error
+
+            As you can see, we need to force the m0 to Shared, and m1 needn't
+            be constrained. The result is always Shared.
+        *)
+        let occ = Maybe_shared.extract_occurrence l1 in
+        Maybe_unique.force_shared_multiuse l0 ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared _, Shared _ -> m0
+    | Maybe_unique l, Shared occ ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:First;
+        Shared occ
+    | Shared occ, Maybe_unique l ->
+        Maybe_unique.force_shared_multiuse l ~there:occ ~first_or_second:Second;
+        Shared occ
+    | Maybe_unique l0, Maybe_unique l1 ->
+        let occ1 = Maybe_unique.extract_occurrence l1 in
+        let occ0 = Maybe_unique.extract_occurrence l0 in
+        Maybe_unique.force_shared_multiuse l0 ~there:occ1 ~first_or_second:First;
+        Maybe_unique.force_shared_multiuse l1 ~there:occ0
+          ~first_or_second:Second;
+        Shared (Maybe_unique.extract_occurrence l0)
+end
+
+(** lifting module Usage to trees *)
+module Usage_tree : sig
+  module Path : sig
+    type t
+    (** Represents a path from the root to a node in a tree *)
+
+    val child : t -> Projection.t -> t
+    (** Constructing a child path *)
+
+    val root : t
+    (** The path representing the root node *)
+  end
+
+  type t
+  (** Usage tree, lifted from [Usage.t] *)
+
+  val print : Format.formatter -> t -> unit
+
+  val seq : t -> t -> t
+  (** Sequential composition lifted from [Usage.seq] *)
+
+  val par : t -> t -> t
+  (** Parallel composition lifted from [Usage.par]  *)
+
+  val singleton : Path.t -> Usage.t -> t
+  (** A singleton tree containing only one leaf *)
+
+  val map : (Path.t -> Usage.t -> Usage.t) -> t -> t
```

Comment:
I think this should have the comment about `map` only working on monotone functions.